### PR TITLE
refactor: enable Ivy and fix template type checking issues

### LINKED ISSUE #16606: stepper: Type 'FormGroup' is not assignable to type 'FormControlLike'
Is it normal?
image

### PR DESCRIPTION
**Pull request by Splaktar**
## Changes
- enable Ivy and fix template type checking in Bazel
  - for libraries, examples, dev-app, CI
  - add postinstall task for ivy-ngcc and configure it for our SystemJS setup
- pin to Angular `9.0.0-next.10` for this PR
  - Angular `9.0.0-next.11` brings in a lot of new template type checking errors
- fix buggy show bounding box checkbox in connected-overlay-demo
- fix stylelint to work with WebStorm plugin
- restrict NodeJS version to 10.x
  - We can't go to NodeJS 12.x without dropping or upgrading Gulp to 4.x

## API Changes
- export `ProgressBarMode` type from `mdc-progress-bar` and `mat-progress-bar`
- export `MatDrawerMode` type from `mat-drawer`
- Add `| null` to `mat-tab-body`'s `@Input() origin: number;`

## TODO
- [x] Remove need for `any` on `@Input() stepControl: FormControlLike | any;` in `CdkStepper`: https://github.com/angular/components/issues/16606
- [x] Figure out how to resolve "Property '`focused`' does not exist on type '`HTMLInputElement`'." when binding to the `focused` property of a `matInput` through a template variable.
- [x] Figure out how to bind `value` field from template variables
- [x] remove `(observableX | async)!` workaround after `cdkVirtualForOf` is updated with better typing like `ngForOf`: https://github.com/angular/components/issues/17411. Fixed in PR https://github.com/angular/components/pull/17421.
- [x] Figure out how to load ngModules in the browser after a successful Ivy compilation. Related issue: https://github.com/angular/angular/issues/31314#issuecomment-542447965
- [ ] Figure out how to bind to `[value]` on a datepicker input w/o the need for `any`. This may be a conflict/bug with template type checking. There is some [discussion of this here](https://github.com/angular/components/pull/16373/files#r296934263).
- [ ] Determine what to do about the hundreds of template type checking errors for boolean and other `@Input`s caused by Angular `9.0.0-next.11`. Related discussion: https://github.com/angular/angular/pull/33066#issuecomment-542458251 and FW-1475.

### Notes
- Settting `"fullTemplateTypeCheck": false` for dev-app doesn't remove the need for many of the changes in this PR. `false` here means, do partial template type checking. It doesn't turn off template type checking.
  - Similarly `"fullTemplateTypeCheck": false` does not resolve the hundreds of errors from Angular `9.0.0-next.11`.

### Other related issues
- `QueryList` is not iterable and can't be used directly in `*ngFor` with strict template type checking: https://github.com/angular/angular/issues/29842

Fixes #16606. Closes #16373.

### CODE REVIEW

**Comment by Splaktar** [comment]
It looks like the issue with the page not loading due to moduleDef undefined errors, may be related to this
```
 --create-ivy-entry-points  If specified then new `*_ivy_ngcc` entry-points
                             will be added to package.json rather than modifying
                             the ones in-place.
                             For this to work you need to have custom resolution
                             set up (e.g. in webpack) to look for these new
                             entry-points.
                             The Angular CLI does this already, so it is safe to
                             use this option if the project is being built via
                             the CLI.                
```
We haven't set up any custom resolution in our SystemJS configuration AFAIK.

File: package.json [hunk 1]
```diff
@@ -9,11 +9,11 @@
   },
   "license": "MIT",
   "engines": {
-    "node": ">= 5.4.1",
+    "node": "10",
     "yarn": ">= 1.17.3"
   },
   "scripts": {
-    "postinstall": "node --preserve-symlinks --preserve-symlinks-main tools/bazel/postinstall-patches.js",
+    "postinstall": "node --preserve-symlinks --preserve-symlinks-main tools/bazel/postinstall-patches.js | ivy-ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points",
```
Comment (devversion) [inline package.json]: I think the issue for the bullet point where the `NgModule`s cannot be loaded in the browser is that we rely in our dev-app on UMD bundles. This is not a bad thing as we do not have tree-shaking anyway for our dev-app.

So, I think the solution would be to process UMD here is as well. I guess the `--first-only` flag would need to be removed. I'm also thinking what stuff we need. Since we only use the Angular framework packages, we should only need `umd` format and the typings transformed. Also I don't think we benefit from the `--create-ivy-entry-points` flag (it rather slows us down)

Comment (Splaktar) [reply package.json]: And `--create-ivy-entry-points` does more than slow us down, it's actually the cause of the page not loading and throwing an exception. It appears that removing this and having it create the Ivy entry points in place, rather than creating new entry points and adding them to `package.json`, is what we need for our SystemJS implementation (dev-app).

Comment (devversion) [reply package.json]: Yeah, we just rely on the original UMD bundles and not look for them in the new ivy ngcc folder.

File: .bazelrc [hunk 2]
```diff
@@ -40,9 +40,8 @@ build:release --workspace_status_command="node ./tools/bazel-stamp-vars.js"
 # Temporary Settings for Ivy   #
 ################################
 
-# Use the legacy AOT compiler strategy. We don't want to compile with Ivy nor with "ngtsc" which
-# does not generate factory files which are needed for AOT.
-build --define=compile=legacy
+# Use the Ivy AOT compiler strategy. We want to compile with Ivy and with "ngtsc".
+build --define=compile=aot --keep_going=true
```
Comment (jelbourn) [inline .bazelrc]: What are the required steps if someone wants to test something in ViewEngine?

File: package.json [hunk 3]
```diff
@@ -9,11 +9,11 @@
   },
   "license": "MIT",
   "engines": {
-    "node": ">= 5.4.1",
+    "node": "10",
```
Comment (crisbeto) [inline package.json]: Shouldn't this be `>= 10`? This won't allow Node 11.

File: package.json [hunk 4]
```diff
@@ -41,33 +41,33 @@
...
-    "@angular/elements": "^9.0.0-next.10",
-    "@angular/forms": "^9.0.0-next.10",
-    "@angular/platform-browser": "^9.0.0-next.10",
+    "@angular/animations": "9.0.0-next.10",
+    "@angular/common": "9.0.0-next.10",
+    "@angular/compiler": "9.0.0-next.10",
+    "@angular/core": "9.0.0-next.10",
+    "@angular/elements": "9.0.0-next.10",
+    "@angular/forms": "9.0.0-next.10",
+    "@angular/platform-browser": "9.0.0-next.10",
```
Comment (crisbeto) [inline package.json]: Might as well bump to next.11 which got release yesterday.

File: src/a11y-demo/checkbox/checkbox-a11y.ts [hunk 5]
```diff
@@ -50,12 +50,18 @@ export class CheckboxAccessibilityDemo {
     return task.completed || (subtasks != null && subtasks.every(t => t.completed));
   }
 
-  someComplete(tasks: Task[]): boolean {
+  someComplete(tasks: Task[] | undefined | null): boolean {
+    if (tasks === undefined || tasks === null) {
```
Comment (crisbeto) [inline src/a11y-demo/checkbox/checkbox-a11y.ts]: You can reduce this to `tasks == null`. Same goes for the one below.

File: src/dev-app/datepicker/datepicker-demo.ts [hunk 6]
```diff
@@ -42,15 +42,20 @@ export class DatepickerDemo {
...
   lastDateChange: Date | null;
   color: ThemePalette;
 
   dateCtrl = new FormControl();
 
-  dateFilter =
-      (date: Date) => !(date.getFullYear() % 2) && (date.getMonth() % 2) && !(date.getDate() % 2)
+  dateFilter: (date: Date | null) => boolean =
+    (date: Date | null) => {
+      if (date === null) {
```
Comment (crisbeto) [inline src/dev-app/datepicker/datepicker-demo.ts]: Maybe just change the one below to `!date || (...)`.

File: src/material/tabs/tab-group.html [hunk 7]
```diff
@@ -1,10 +1,10 @@
 <mat-tab-header #tabHeader
-               [selectedIndex]="selectedIndex"
+               [selectedIndex]="selectedIndex || 0"
                [disableRipple]="disableRipple"
                (indexFocused)="_focusChanged($event)"
                (selectFocusedIndex)="selectedIndex = $event">
   <div class="mat-tab-label" role="tab" matTabLabelWrapper mat-ripple cdkMonitorElementFocus
-       *ngFor="let tab of _tabs; let i = index"
+       *ngFor="let tab of _tabs.toArray(); let i = index"
```
Comment (crisbeto) [inline src/material/tabs/tab-group.html]: This isn't very efficient, because `toArray` returns a new array on each change detection. If `ngFor` requires an array, we should have a different property that we only re-assign when the tabs change.

File: .bazelrc [hunk 8]
```diff
@@ -40,9 +40,8 @@ build:release --workspace_status_command="node ./tools/bazel-stamp-vars.js"
 # Temporary Settings for Ivy   #
 ################################
 
-# Use the legacy AOT compiler strategy. We don't want to compile with Ivy nor with "ngtsc" which
-# does not generate factory files which are needed for AOT.
-build --define=compile=legacy
+# Use the Ivy AOT compiler strategy. We want to compile with Ivy and with "ngtsc".
```
Comment (devversion) [inline .bazelrc]: There Ivy compiler is always AOT. So just say that we want to compile with `ngtsc`.

File: .bazelrc (hunk 2 above)
Comment (devversion) [inline .bazelrc]: Why add `--keep_going=true` here? This should _not_ be part of the default bazelrc. Especially not of this section which is primarily about the compile mode for `ng_module.bzl`

File: .circleci/config.yml [hunk 9]
```diff
@@ -433,10 +429,6 @@ jobs:
 
       # Setup Angular ivy snapshots built with ngtsc.
       - run: node ./scripts/circleci/setup-angular-snapshots.js --tag master-ivy-aot
-      # Disable type checking when building with Ivy. This is necessary because
-      # type checking is not complete yet and can incorrectly break compilation.
-      # Issue is tracked with FW-1004.
-      - run: sed -i "s/\(_ENABLE_NG_TYPE_CHECKING = \)True/\1False/g" tools/defaults.bzl
```
Comment (devversion) [inline .circleci/config.yml]: Can you remove the `ENABLE_NG_TYPE_CHECKING` flag from `defaults.bzl` too? We just had that so that we can disable type checking here?

File: package.json [hunk 10]
```diff
@@ -9,11 +9,11 @@
   },
   "license": "MIT",
   "engines": {
-    "node": ">= 5.4.1",
+    "node": "10",
     "yarn": ">= 1.17.3"
   },
   "scripts": {
-    "postinstall": "node --preserve-symlinks --preserve-symlinks-main tools/bazel/postinstall-patches.js",
+    "postinstall": "node --preserve-symlinks --preserve-symlinks-main tools/bazel/postinstall-patches.js | ivy-ngcc --properties main",
```
Comment (devversion) [inline package.json]: It might be good to only do this when running within Bazel. ngcc is not very fast and otherwise we are running ngcc twice. We can tackle this in a follow-up though.

File: src/dev-app/checkbox/checkbox-demo.ts [hunk 11]
```diff
@@ -46,7 +47,7 @@ export class AnimationsNoop {
       margin-bottom: 4px;
     }
   `],
-  templateUrl: 'nested-checklist.html',
+  templateUrl: './nested-checklist.html',
```
Comment (devversion) [inline src/dev-app/checkbox/checkbox-demo.ts]: why is that needed? (just asking out of curiosity; anything that changed in ngtsc that needs this?)

File: src/dev-app/connected-overlay/connected-overlay-demo.ts [hunk 12]
```diff
@@ -92,16 +99,19 @@ export class ConnectedOverlayDemo {
...
     const box = document.querySelector<HTMLElement>('.cdk-overlay-connected-position-bounding-box');
 
     if (box) {
-      this.showBoundingBox = !this.showBoundingBox;
-      box.style.background = this.showBoundingBox ? 'rgb(255, 69, 0, 0.2)' : '';
+      if (showBoundingBox) {
+        box.classList.add('demo-bounding-box-visible');
+      } else {
+        box.classList.remove('demo-bounding-box-visible');
+      }
```
Comment (devversion) [inline src/dev-app/connected-overlay/connected-overlay-demo.ts]: Why has this been updated? I'm asking because I'd expect only changes which are relevant to Ivy / type checking.

File: src/dev-app/datepicker/datepicker-demo.html [hunk 13]
```diff
@@ -138,8 +138,13 @@ <h2>Datepicker with value property binding</h2>
   <mat-datepicker-toggle [for]="datePicker4"></mat-datepicker-toggle>
   <mat-form-field>
     <mat-label>Value binding</mat-label>
+    <!--  TODO Angular sets the inputs for both directives on the input element here.
+          e.g. [value] is set for MatInput and MatDatepicker.
+          MatInput#value expects a string whereas MatDatepicker#value expects a Date.
+          This breaks strict template type checking. What should we do here?
+    -->
```
Comment (devversion) [inline src/dev-app/datepicker/datepicker-demo.html]: Re-opening a comment here again ([we had one in the other PR for this](https://github.com/angular/components/pull/16373#discussion_r296934263)). I'm not sure if anything has changed here. Not even sure if we have a ticket for it. I don't think we talked about it in the framework meeting yet.

**Comment by devversion** [comment]
Also we should update the PR/commit message. This PR is actually turning on Ivy for the whole repository.

File: package.json (hunk 3 above)
Comment (Splaktar) [reply package.json]: NodeJS 11 went [EOL](https://nodejs.org/en/about/releases/) a number of months ago. It's not supported or recommended. It can cause bugs and other issues with a number of libraries.

File: .bazelrc (hunk 2 above)
Comment (Splaktar) [reply .bazelrc]: It's just for debugging and fixing in this branch. I wasn't planning to leave it. Is there a location that it would make sense to put a toggle for it or a better place to leave a comment about where it should be enabled when needed?

File: package.json (hunk 4 above)
Comment (Splaktar) [reply package.json]: As mentioned in the OP, Angular `9.0.0-next.11` breaks "everything", i.e. hundreds of type checking errors, even with `"fullTemplateTypeCheck": false`. I've raised the issue with Joost and Jeremy has discussed it with Kara and Alex R. Hopefully `next.12` will be better?

File: src/dev-app/checkbox/checkbox-demo.ts (hunk 11 above)
Comment (Splaktar) [reply src/dev-app/checkbox/checkbox-demo.ts]: I'm not really sure why, but without this, WebStorm was unable to resolve the class functions and fields in the template.

It only seemed to be a problem with this one. If it's problematic, I can revert this.

File: src/dev-app/connected-overlay/connected-overlay-demo.ts (hunk 12 above)
Comment (Splaktar) [reply src/dev-app/connected-overlay/connected-overlay-demo.ts]: I ran into this while verifying the components via the dev-app pages. Previously if you clicked the checkbox, it would check itself and then the previous code here would instantly uncheck the box. Then it you toggled it off again the behavior would be the opposite of what was expected. I fixed it to verify that it wasn't an Ivy bug, which it wasn't, just a buggy demo.

File: src/dev-app/datepicker/datepicker-demo.html (hunk 13 above)
Comment (Splaktar) [reply src/dev-app/datepicker/datepicker-demo.html]: Jeremy spoke to Alex R about it. It sounds like this is a real issue and that Alex was going to investigate a fix. We should probably open a ticket in JIRA?

File: src/dev-app/datepicker/datepicker-demo.ts (hunk 6 above)
Comment (Splaktar) [reply src/dev-app/datepicker/datepicker-demo.ts]: I considered what Paul had in https://github.com/angular/components/pull/16373/files#diff-72d99d967518ce37f1e422881c29a43aR60, but it seemed to be less readable.

Your suggestion is?
```ts
return !date || !(date.getFullYear() % 2) && Boolean(date.getMonth() % 2) && !(date.getDate() % 2);
```

File: src/material/tabs/tab-group.html (hunk 7 above)
Comment (Splaktar) [reply src/material/tabs/tab-group.html]: What should we call the field? `_tabsArray`? `_tabsIterable`?

File: .bazelrc (hunk 2 above)
Comment (Splaktar) [reply .bazelrc]: To build for ViewEngine, you can use `--define=compile=legacy` here instead. I think that @devversion has some ideas on how to make this more configurable and hopefully add NPM scripts to the `package.json` for building with Ivy (default) or ViewEngine.

I think that the deploy scripts will need to be changed to always build with ViewEngine since, in Angular version 9, that's the only way to publish libraries.

File: src/dev-app/datepicker/datepicker-demo.ts (hunk 6 above)
Comment (crisbeto) [reply src/dev-app/datepicker/datepicker-demo.ts]: Yes, exactly. It's a nit so feel free to ignore.

File: src/material/tabs/tab-group.html (hunk 7 above)
Comment (crisbeto) [reply src/material/tabs/tab-group.html]: I'd say `_tabsArray`.

Comment (Splaktar) [reply src/material/tabs/tab-group.html]: Done for Tabs and Stepper.

**Comment by Splaktar** [comment]
I've got a WIP PR open that contains these changes plus https://github.com/angular/components/commit/731b80de08f0495481822ff713f9365df3b1e4bf which starts to resolve all of the `next.11` template type checking errors with `"fullTemplateTypeCheck": true,`.

**Comment by Splaktar** [comment]
It looks like `@angular/schematics@9.0.0-next.12` requires TypeScript `3.6`. I'm reverting to version of the Angular CLI related libraries to resolve this.

**Comment by Splaktar** [comment]
I had to revert Angular `9.0.0-next.11` as well as it runs into strange issues with `ivy_test` failing due to npm or yarn errors.

**Comment by Splaktar** [comment]
OK, hopefully this last push resolves the issues with the `public_api_guard` and updates the goldens so that everything is finally green.

File: tools/defaults.bzl [hunk 14]
```diff
@@ -97,7 +92,7 @@ def ng_module(
 
     _ng_module(
         srcs = srcs,
-        type_check = _ENABLE_NG_TYPE_CHECKING,
+        type_check = True,
```
Comment (devversion) [inline tools/defaults.bzl]: We can remove this, since it's enabled by default.

File: src/dev-app/connected-overlay/connected-overlay-demo.ts (hunk 12 above)
Comment (devversion) [reply src/dev-app/connected-overlay/connected-overlay-demo.ts]: Sounds good. Just checking so that we don't oversee any potential breaking changes.

File: tools/defaults.bzl (hunk 14 above)
Comment (Splaktar) [reply tools/defaults.bzl]: Done
